codex/sales/reservations.nim

```diff
@@ -217,144 +300,184 @@ proc reserve*(
-    return failure(err)
+    trace "rolling back reservation creation"
 
-  return success()
+    without key =? reservation.key, keyError:
+      keyError.parent = updateErr
+      return failure(keyError)
+
+    # rollback the reservation creation
+    if rollbackErr =? (await self.delete(key)).errorOption:
+      rollbackErr.parent = updateErr
+      return failure(rollbackErr)
 
+    return failure(updateErr)
 
-proc markUsed*(
+  return success(reservation)
+
+proc release*(
   self: Reservations,
-  id: AvailabilityId): Future[?!void] {.async.} =
+  reservationId: ReservationId,
+  availabilityId: AvailabilityId,
+  bytes: uint): Future[?!void] {.async.} =
 
-  without var availability =? (await self.get(id)), err:
-    return failure(err)
+  logScope:
+    topics = "release"
+    bytes
+    reservationId
+    availabilityId
 
-  availability.used = true
-  let r = await self.update(availability)
-  if r.isOk:
-    trace "availability marked used", id = id.toArray.toHex
-  return r
+  trace "releasing bytes and updating reservation"
 
-proc markUnused*(
-  self: Reservations,
-  id: AvailabilityId): Future[?!void] {.async.} =
+  without key =? key(reservationId, availabilityId), error:
+    return failure(error)
+
+  without var reservation =? (await self.get(key, Reservation)), error:
+    return failure(error)
 
-  without var availability =? (await self.get(id)), err:
+  if reservation.size < bytes.u256:
+    let error = newException(BytesOutOfBoundsError,
+      "trying to release an amount of bytes that is greater than the total " &
+      "size of the Reservation")
+    return failure(error)
+
+  if releaseErr =? (await self.repo.release(bytes)).errorOption:
+    return failure(releaseErr.toErr(ReleaseFailedError))
+
+  reservation.size -= bytes.u256
+
+  # persist partially used Reservation with updated size
+  if err =? (await self.update(reservation)).errorOption:
+
+    # rollback release if an update error encountered
+    trace "rolling back release"
+    if rollbackErr =? (await self.repo.reserve(bytes)).errorOption:
+      rollbackErr.parent = err
+      return failure(rollbackErr)
     return failure(err)
 
-  availability.used = false
-  let r = await self.update(availability)
-  if r.isOk:
-    trace "availability marked unused", id = id.toArray.toHex
-  return r
+  return success()
 
-iterator items*(self: AvailabilityIter): Future[?Availability] =
+iterator items(self: StorableIter): Future[?seq[byte]] =
   while not self.finished:
     yield self.next()
 
-proc availabilities*(
-  self: Reservations): Future[?!AvailabilityIter] {.async.} =
+proc storables(
+  self: Reservations,
+  T: type SomeStorableObject
+): Future[?!StorableIter] {.async.} =
 
-  var iter = AvailabilityIter()
+  var iter = StorableIter()
   let query = Query.init(ReservationsKey)
-
-  without results =? await self.repo.metaDs.query(query), err:
-    return failure(err)
-
-  proc next(): Future[?Availability] {.async.} =
+  when T is Availability:
```

Comment:
Wouldn't it be nicer to have two procs here, one for Availability, and one for Reservation? Now the compiler won't help you if you were to introduce another SomeStorableObject type later.

Reply:
I partially agree, but I decided against it simply because it is not a small proc and it would be a lot of repetition. I opted instead to include both `Availability` and `Reservation` type checks in the `when`, and added an `else` condition that will fail when the type is neither of those two. That way there will be compiler support if a new type is added to `SomeStorableObject`.

Reply:
Updated